QueryBuilder now supports `Traversable` objects for `IN` clauses.

SamMousa opened the pull request.

SamMousa committed:
QueryBuilder now supports `Traversable` objects for `IN` clauses.

SilverFire commented:
Changelog line, please

SamMousa commented:
You'd think I'd learn... :8ball:

SamMousa committed:
Added changelog line.

SamMousa commented:
Implements #10954

SamMousa committed:
Merge remote-tracking branch 'upstream/master' into enh-10954-querybuilder-traversable

cebe commented on framework/db/QueryBuilder.php:
```diff
@@ -1112,7 +1113,7 @@ public function buildInCondition($operator, $operands, &$params)
         }
 
-        $values = (array) $values;
+        $values = ArrayHelper::toArray($values);
```
should set recursive parameter to false to avoid complete iteration over the values if it is already an array.

cebe commented on framework/db/QueryBuilder.php:
```diff
@@ -1202,13 +1203,13 @@ protected function buildCompositeInCondition($operator, $columns, $values, &$par
-            if (strpos($column, '(') === false) {
-                $columns[$i] = $this->db->quoteColumnName($column);
-            }
+            $sqlColumns[] = strpos($column, '(') === false ? $this->db->quoteColumnName($column) : $column;
```
why did you change this?

SamMousa replied:
Ah, my intention was to not call `ArrayHelper::toArray()` and only iterate the values once regardless of it being an array or `\Traversable`.

cebe replied:
Oh, now I see. makes sense.

SamMousa commented:
Still working on some improvements; while reading the code I found that `buildCompositeInCondition` does not support db expressions; is there a reason for that?

cebe commented:
I do not think there is a specific reason, `Expression` support has been added partially over time to more and more methods.

SamMousa committed:
Added support & tests for (composite) in conditions using `\Traversable` parameters.

SamMousa commented:
Ok, my implementation is ready so feedback is welcome :)

SamMousa committed:
Added `TraversableObject` to aid in testing.

SamMousa commented:
Okay some tests that I've added are failing and I'm not sure what the best way to fix them is.

Basically `sqlite` and `mssql` drivers do not support composite in conditions like:
```` where (id, name) in ((1, "Bob"), (2, "Jane"))````

Previously there was no test case for this, but I've added one. The new test fails because there is 1 dataprovider that is used for all query builders, but not all query builders have the same expected value.

Should I instead move these tests to a different test function and not use the dataprovider?

SamMousa committed:
Removed some debug leftover.

SilverFire commented:
I remember about your PR, but do not have enough time to review it right now

SilverFire commented:
Could you updated from master?

SamMousa commented:
Yes I will, but regardless tests will fail until I know how to solve that ;-)

samdark commented:
Yes, these test should be driver-specific.

SilverFire commented:
Fixed tests and merged. Thank you @SamMousa :beers: 
https://github.com/yiisoft/yii2/commit/dcf42e35cc114a8a4652140c0c6d198eaeac2a66
